File: 124. Binary Tree Maximum Path Sum/maxPathSum.js
// Definition for a binary tree node.
class TreeNode {
  constructor(val = 0, left = null, right = null) {
    this.val = val;
    this.left = left;
    this.right = right;
  }
}
/**
 * @param {TreeNode} root
 * @return {number}
 */
const maxPathSum = (root) => {
  let res = root.val;

  const dfs = (node) => {
    if (!node) {
      return 0;
    }

    const leftMax = Math.max(dfs(node.left), 0);
    const rightMax = Math.max(dfs(node.right), 0);

    res = Math.max(res, node.val + leftMax + rightMax);

    return node.val + Math.max(leftMax, rightMax);
  };

  dfs(root);
  return res;
};


const root = new TreeNode(-10);
root.left = new TreeNode(9);
root.right = new TreeNode(20);

root.right.left = new TreeNode(15);
root.right.right = new TreeNode(7);

console.log('maxPathSum: ', maxPathSum(root));

File: 124. Binary Tree Maximum Path Sum/maxPathSum.py
from typing import Optional

# Definition for a binary tree node.
class TreeNode:
  def __init__(self, val=0, left=None, right=None):
    self.val = val
    self.left = left
    self.right = right

class Solution:
  def maxPathSum(self, root: Optional[TreeNode]) -> int:
    res = [root.val]
   
    def dfs(node):
      if not node:
        return 0
      
      leftMax = max(dfs(node.left), 0)
      rightMax = max(dfs(node.right), 0)
      
      res[0] = max(res[0], node.val + leftMax + rightMax)
      
      return node.val + max(leftMax, rightMax)
    
    dfs(root)
    return res[0]


root = TreeNode(-10)
root.left = TreeNode(9)
root.right = TreeNode(20)

root.right.left = TreeNode(15)
root.right.right = TreeNode(7)

solution = Solution()
print('maxPathSum: ', solution.maxPathSum(root))
